pkg/canary/hpa_reconciler.go

```diff
@@ -0,0 +1,258 @@
+		if betaErr != nil {
+			return fmt.Errorf("HorizontalPodAutoscaler %s.%s get query error for both v2beta2: %s and v2: %s",
+				cd.Spec.AutoscalerRef.Name, cd.Namespace, betaErr, err)
+		}
+	}
+
+	if hpa != nil {
+		if err = hr.reconcilePrimaryHpaV2(cd, hpa, init); err != nil {
+			return err
+		}
+	} else if betaHpa != nil {
+		if err = hr.reconcilePrimaryHpaV2Beta2(cd, betaHpa, init); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (hr *HPAReconciler) reconcilePrimaryHpaV2(cd *flaggerv1.Canary, hpa *hpav2.HorizontalPodAutoscaler, init bool) error {
+	primaryName := fmt.Sprintf("%s-primary", cd.Spec.TargetRef.Name)
+
+	hpaSpec := hpav2.HorizontalPodAutoscalerSpec{
+		ScaleTargetRef: hpav2.CrossVersionObjectReference{
+			Name:       primaryName,
+			Kind:       hpa.Spec.ScaleTargetRef.Kind,
+			APIVersion: hpa.Spec.ScaleTargetRef.APIVersion,
+		},
+		MinReplicas: hpa.Spec.MinReplicas,
+		MaxReplicas: hpa.Spec.MaxReplicas,
+		Metrics:     hpa.Spec.Metrics,
+		Behavior:    hpa.Spec.Behavior,
+	}
+
+	primaryHpaName := fmt.Sprintf("%s-primary", cd.Spec.AutoscalerRef.Name)
+	primaryHpa, err := hr.kubeClient.AutoscalingV2().HorizontalPodAutoscalers(cd.Namespace).Get(context.TODO(), primaryHpaName, metav1.GetOptions{})
+
+	// create HPA
+	if errors.IsNotFound(err) {
+		primaryHpa = &hpav2.HorizontalPodAutoscaler{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      primaryHpaName,
+				Namespace: cd.Namespace,
+				Labels:    filterMetadata(hpa.Labels),
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(cd, schema.GroupVersionKind{
+						Group:   flaggerv1.SchemeGroupVersion.Group,
+						Version: flaggerv1.SchemeGroupVersion.Version,
+						Kind:    flaggerv1.CanaryKind,
+					}),
+				},
+			},
+			Spec: hpaSpec,
+		}
+
+		_, err = hr.kubeClient.AutoscalingV2().HorizontalPodAutoscalers(cd.Namespace).Create(context.TODO(), primaryHpa, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("creating HorizontalPodAutoscaler %s.%s failed: %w",
+				primaryHpa.Name, primaryHpa.Namespace, err)
+		}
+		hr.logger.With("canary", fmt.Sprintf("%s.%s", cd.Name, cd.Namespace)).Infof(
+			"HorizontalPodAutoscaler %s.%s created", primaryHpa.GetName(), cd.Namespace)
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("HorizontalPodAutoscaler %s.%s get query failed: %w",
+			primaryHpa.Name, primaryHpa.Namespace, err)
+	}
+
+	// update HPA
+	if !init && primaryHpa != nil {
+		diffMetrics := cmp.Diff(hpaSpec.Metrics, primaryHpa.Spec.Metrics)
+		diffBehavior := cmp.Diff(hpaSpec.Behavior, primaryHpa.Spec.Behavior)
+		diffLabels := cmp.Diff(hpa.ObjectMeta.Labels, primaryHpa.ObjectMeta.Labels)
+		diffAnnotations := cmp.Diff(hpa.ObjectMeta.Annotations, primaryHpa.ObjectMeta.Annotations)
+		if diffMetrics != "" || diffBehavior != "" || diffLabels != "" || diffAnnotations != "" || int32Default(hpaSpec.MinReplicas) != int32Default(primaryHpa.Spec.MinReplicas) || hpaSpec.MaxReplicas != primaryHpa.Spec.MaxReplicas {
```

Comment:
Let's extract this to a function like `hasChanged` and reuse it for both v2beta2 and v2.